src/client.rs

```diff
@@ -429,54 +400,57 @@ impl Client {
-                || q_lower.starts_with("show")
-            {
-                self.client.get(url).send()?
-            } else {
-                self.client.post(url).send()?
-            }
+        let resp_future = if q_lower.starts_with("select") && !q_lower.contains("into")
+            || q_lower.starts_with("show")
+        {
+            self.client.get(url).send().boxed()
+        } else {
+            self.client.post(url).send().boxed()
         };
 
-        match res.status_raw().0 {
-            200 => Ok(res),
-            400 => {
-                let mut context = String::new();
-                let _ = res.read_to_string(&mut context);
-                let json_data: Query = serde_json::from_str(&context).unwrap();
-
-                Err(error::Error::SyntaxError(serialization::conversion(
-                    &json_data.error.unwrap(),
-                )))
+        async move {
+            let res = resp_future.await?;
+            match res.status().as_u16() {
+                200 => Ok(res),
+                400 => {
+                    let json_data: Query = res.json().await?;
+
+                    Err(error::Error::SyntaxError(serialization::conversion(
+                        &json_data.error.unwrap(),
+                    )))
+                }
+                401 | 403 => Err(error::Error::InvalidCredentials(
+                    "Invalid authentication credentials.".to_string(),
+                )),
+                _ => Err(error::Error::Unknow("There is something wrong".to_string())),
             }
-            401 | 403 => Err(error::Error::InvalidCredentials(
-                "Invalid authentication credentials.".to_string(),
-            )),
-            _ => Err(error::Error::Unknow("There is something wrong".to_string())),
         }
     }
 
     /// Query and return to the native json structure
-    fn query_raw(&self, q: &str, epoch: Option<Precision>) -> Result<Query, error::Error> {
-        let mut response = self.send_request(q, epoch, false)?;
-
-        let mut context = String::new();
-        let _ = response.read_to_string(&mut context);
-
-        let json_data: Query = serde_json::from_str(&context).unwrap();
-        Ok(json_data)
+    fn query_raw(
+        &self,
+        q: &str,
+        epoch: Option<Precision>,
+    ) -> impl Future<Output = Result<Query, error::Error>> {
+        let resp_future = self.send_request(q, epoch, false);
+        async move { Ok(resp_future.await?.json().await?) }
     }
 
     /// Query and return to the native json structure
     fn query_raw_chunked(
         &self,
         q: &str,
         epoch: Option<Precision>,
-    ) -> Result<ChunkedQuery<SerdeIoRead<Response>>, error::Error> {
-        let response = self.send_request(q, epoch, true)?;
-        let stream = serde_json::Deserializer::from_reader(response).into_iter::<Query>();
-        Ok(stream)
+    ) -> impl Future<Output = Result<ChunkedQuery<'static, IoRead<Cursor<Bytes>>>, error::Error>>
```

Comment:
Why the type is bytes? Looks like something wrong with this type

Reply:
`Cursor<Bytes>` is the type we are deserializing from.

``` rust
pub type ChunkedQuery<'de, T> = serde_json::StreamDeserializer<'de, T, Query>;
```

`Query` is the type we are deserializing to. 

It is a `Bytes` because reqwest uses that instead of `Vec<u8>`.

Reply:
I think we should remove `ChunkedQuery` and return `Result<Query>` directly, what do you think?

Reply:
> I think we should remove `ChunkedQuery` and return `Result<Query>` directly, what do you think?

This is not a good idea. The meaning of chunk here is that you can lazy parse the response like iterator.

Reply:
This is a valid use-case, however, currently it is less meaningful as we always have to collect the entire buffer, because serde doesn't support reading from AsyncRead.

Relevant issue: https://github.com/serde-rs/json/issues/575

Reply:
If would be best if we could use [bytes_stream](https://docs.rs/reqwest/0.10.3/reqwest/struct.Response.html#method.bytes_stream), I'm not sure how to use it with serde.

Reply:
> If would be best if we could use [bytes_stream](https://docs.rs/reqwest/0.10.3/reqwest/struct.Response.html#method.bytes_stream), I'm not sure how to use it with serde.

I think it may like 
```
stream = reqwest::get("http://httpbin.org/ip")
    .await?
    .bytes_stream().and_then(|bytes| serde_json::Deserializer::from_reader(bytes))
```

it will return `impl Future<output = Result<Query>>`, and it also a lazy iterator

Reply:
It won't work though, because one json value can be split into multiple bytes.